### PR TITLE
chain sub-service exceptions, clean up error reporting

### PR DESCRIPTION
**Pull request by tedsharpe**
Addresses DSDEEP-1338.

- [x] **Submitter**: Rebase to master. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] **Submitter**: **Assign PR to** 
- [x] : does a once-over of the PR
- [x] : Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  * Rest of team may comments on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter updates documentation as needed
  * Submitter rebases to master again if necessary
  * Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x] **LR**: sign off, **assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to master
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: Verify swagger UI on dev server still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -254,7 +254,7 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, containerDAO:
     }
 
     // collect the result
-    eventualResponses map (r => RequestComplete(r.collect { case Some(x) => x }))
+    eventualResponses map (r => RequestComplete(StatusCodes.OK, r.collect { case Some(x) => x }))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Is it required to specify status code now?

**Comment by dvoet** [comment]
:+1: 

can you write a paragraph about how we should use ErrorReport going forward?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: It isn't required, but I think you might find it convenient when writing unit tests directly against the WorkspaceService methods:  you can rely on 2XX responses always returning a duple RequestComplete[(StatusCode, Any)], and all other responses returning RequestComplete[ErrorReport].  But the old, random way is still supported.

**Comment by tedsharpe** [comment]
My suggestion is that going forward, you should always return a RequestComplete[ErrorReport] whenever the status code you'll be returning is not in the 2XX series.

At the top level, in WorkspaceService, you can just use ErrorReport(statusCode: StatusCode, message: String).  This will tell callers that the source of the error is rawls, and the response won't include a lot of extra drek.  For example, in withEntity, the result when an entity can't be found in OrientDB is:
<pre>RequestComplete(ErrorReport(StatusCodes.NotFound, s"${entityType} ${entityName} does not exist in ${workspaceContext}"))</pre>

More complex cases occur when you have called a sub-service API which resulted in an exception.  You will want to pass back an error report from rawls that includes a nested report attributed to the sub-service that caused it.  In that case you'll want to do something like this example from listAdmins:
<pre>RequestComplete(ErrorReport(StatusCodes.BadGateway,"Unable to list admins.",gcsDAO.toErrorReport(throwable)))<pre>

Most complex of all is when you need to gather many causes together in one error report.  This typically occurs when you have an endpoint that attempts to apply a number of operations.  For example, updateACL may have many googly error reports which have been gathered together into val reports:
<pre>RequestComplete(ErrorReport(StatusCodes.Conflict,"Unable to alter some ACLs in $workspaceName",reports))</pre>

Sub-service DAOs ought to provide a mechanism for turning a failed Future into an ErrorReport attributed to the service that barfed.  (They now all have a toErrorReport(throwable: Throwable) method.)  If you recognize the subtype of Throwable and can extract a StatusCode from it, and figure out where the sub-service stashed the most specific and helpful error message, so much the better.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (tedsharpe) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Actually, that isn't quite true, is it.  There are a few 2XX responses of the shape RequestComplete[StatusCode].  Ideally the status code should be NoContent for all of those, but I know for a fact that it isn't.

**Pull request merged**